tools/tpm2_startauthsession.c

```diff
@@ -272,9 +292,34 @@ static tool_rc process_input_data(ESYS_CONTEXT *ectx) {
 
             bool is_transient = (ctx.session.tpmkey.key_context_object.handle
                     >> TPM2_HR_SHIFT) == TPM2_HT_TRANSIENT;
-            if (!is_transient) {
+            if (!is_transient && !ctx.name_path) {
                 LOG_WARN("check public portion of the tpmkey manually");
             }
+
+            /*
+             * ESAPI performs this check when an ESYS_TR or Context file is used, so we
+             * could only run the check on the case where a raw TPM handle is provided,
+             * however, it seems prudent that if the user specifies a name, we always
+             * just check it.
+             */
+            if (ctx.name_path) {
+                TPM2B_NAME *got_name = NULL;
+                rc = tpm2_tr_get_name(ectx, ctx.session.tpmkey.key_context_object.tr_handle,
+                        &got_name);
+                if (rc != tool_rc_success) {
+                    Esys_Free(got_name);
+                    return rc;
+                }
+
+                if (ctx.name.size != got_name->size ||
+                        memcmp(got_name->name, ctx.name.name, ctx.name.size)) {
+                    LOG_ERR("Expected name does not match");
```

Comment:
```suggestion
                bool is_expected_object = (memcmp(got_name->name, ctx.name.name, ctx.name.size) == 0);
                if (ctx.name.size != got_name->size || !is_expected_object) {
                    LOG_ERR("Expected name does not match");
```

Reply:
> 
You MUST check size before you do memcmp. Size should also be bounded by the buffer size, I wonder how many of those bugs we have floating around. I'll write a compare TPM2B function.